### PR TITLE
Enhance YAML-based testing framework

### PR DESCRIPTION
**Pull request by srikanth-kumar**
# Summary

This PR significantly enhances the YAML-based testing framework and addresses multiple bugs and edge cases: 

-  **Corrects Silent Test Passes** 
    - Ensures tests fail explicitly when `plugin_model(**bad_plugin)` unexpectedly does not raise a validation error. 
- **Fixes Improper Validation Expectations**
  - The `validate_bad_plugin()` function previously always expected a `ValidationError`, even for test cases intended to validate warnings or those involving known unimplemented keys such as `abspath`. 
- **Avoids Redundant Calls** 
  - Eliminates unnecessary calls to `plugin_model(**bad_plugin)`.
- **Improves Modularity** 
  - Extracts test_tup validation into a dedicated utility function, `_validate_test_tup_keys` for cleaner and reusable logic.
- **Fixes ResourceWarnings**
  - Resolves `ResourceWarning`s caused by unclosed file pointer in `geoips/tests/unit_tests/pydantic/utils.py`  and `geoips/tests/unit_tests/pydantic/sectors/test_sectors.py`.

# Reviewer Checklist

PR author: Please ensure you meet all of the below requirements, and check boxes appropriately.

Reviewers: Please confirm all required testing/documentation has been completed prior to approving.

**Remove lines that are not applicable, explain if you select "NO REQUIRED"**

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] Required ***unit tests*** added and pass for new/modified functionality
* [x] NO Required ***integration tests*** added and pass for new/modified functionality
* [x] NO REQUIRED ***documentation*** (explain why not required)
* [x] Required ***release notes*** added for new/modified functionality
* [x] No Required ***updates to other repos*** complete

https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes NRLMMD-GEOIPS/geoips#1016

# Testing Instructions

Run pytest within `~/geoips/geoips/tests/unit_tests/pydantic$` directory.

image

### CODE REVIEW

File: tests/unit_tests/pydantic/sectors/test_cases.yaml [hunk 1]
```diff
@@ -61,12 +61,12 @@ shape_missing_height:
       15000
   cls: SectorShape
   err_str: Field required
-center_missing_y:
-  key: spec/center
-  val:
-    x: 0
-  cls: AreaDefinitionSpec
-  err_str: Field required
+# center_missing_y:
```
Comment (evrose54) [inline tests/unit_tests/pydantic/sectors/test_cases.yaml:64]: Why'd you comment this out?

**Review by evrose54 — DISMISSED** [review]
Looks good to me. Thanks Kumar! I added one comment asking why a code snippet was commented out, but otherwise looks good!

**Review by evrose54 — DISMISSED** [review]
Approved! Thanks a ton for doing this Kumar!

**Review by mindyls — DISMISSED** [review]

**Review by jsolbrig — APPROVED** [review]
